Keystone: Endpoint controller

winiciusallan opened the pull request:
This PR introduces the Keystone Endpoint controller.

Fixes: #592

Reviewer notes:
- At this time, Gophercloud requires the `Name` field on creation, and name is deprecated according to the [docs](https://docs.openstack.org/api-ref/identity/v3/index.html?expanded=create-endpoint-detail#create-endpoint), so since I can't omit it for now, I have decided to implement the controller with this field. I opened a [PR](https://github.com/gophercloud/gophercloud/pull/3581) to address it
- Also, we couldn't set a value for the `enabled` field on creation neither on update, so I have to omit it mainly in the tests. This [PR](https://github.com/gophercloud/gophercloud/pull/3584) was opened to fix it,

mandre commented on internal/controllers/endpoint/tests/endpoint-import-dependency/03-delete-import-dependencies.yaml:
```diff
@@ -3,5 +3,5 @@ apiVersion: kuttl.dev/v1beta1
 commands:
   # We should be able to delete the import dependencies
-  - command: kubectl delete service endpoint-import-dependency
+  - command: kubectl delete services.openstack.k-orc.cloud endpoint-import-dependency
```
We should update the generator to use the fully qualified name for kubectl commands to avoid problems like this. I'll create a PR.

mandre reviewed (commented):
I wasn't sure at first, but I now believe we should wait for the next gophercloud release to avoid the introduction of a backward incompatible change in the API when we drop the Name field.

winiciusallan commented:
Hey, sorry for the delay. Now that we have a new gophercloud release, can we get this one #628?

Once merged, I'll focus on addressing your comments and taking it further.

winiciusallan commented:
@mandre I noticed that I missed a pointer on `Description` in endpoint.UpdateOpts :sweat_smile:, so we are not able to unset that field; tests has been falling on `endpoint-update`. When we pass the zero-value, it omits the field and does not update it =/. I opened a [PR](https://github.com/gophercloud/gophercloud/pull/3599) for it.

Is it okay if we ignore it for now and implement it later (considering that a gophercloud release was launched recently)?

https://github.com/gophercloud/gophercloud/blob/main/openstack/identity/v3/endpoints/requests.go#L137

winiciusallan commented:
@mandre Hey, I'll work on the final adjustments today or tomorrow at the latest. I'll let you know when it'll be ready.

mandre commented on api/v1alpha1/endpoint_types.go:
```diff
@@ -0,0 +1,91 @@
+
+	// enabled indicates whether the endpoint is enabled or not.
+	// +optional
+	Enabled bool `json:"enabled,omitempty"`
```
We want to be able to differentiate between `false` and unset.
```suggestion
	Enabled *bool `json:"enabled,omitempty"`
```

mandre commented on config/samples/openstack_v1alpha1_endpoint.yaml:
```diff
@@ -0,0 +1,20 @@
+  name: endpoint-sample
+spec:
+  cloudCredentialsRef:
+<<<<<<< Updated upstream
```
There's an unresolved merge conflict.

mandre commented on api/v1alpha1/endpoint_types.go:
```diff
@@ -72,17 +69,22 @@ type EndpointResourceStatus struct {
-	// description is a human-readable description for the resource.
+	// enabled indicates whether the endpoint is enabled or not.
+	// +optional
+	Enabled *bool `json:"enabled,omitempty"`
```
My bad, this was bad advice. We actually want the pointer to differentiate between false and unset, like we do for other resources.
